Java: Add/improve insecure trustmanager query

intrigus-lgtm opened the pull request:
This can be rebased once #4771 is merged.

smowton commented:
Waiting for #4771 before reviewing this

kevinbackhouse commented:
Hi @intrigus-lgtm. It looks like #4771 has been merged. Are you able to rebase this now?

intrigus-lgtm commented:
@kevinbackhouse sure! But I don't have time right now, probably later this day.

intrigus-lgtm commented:
@kevinbackhouse @smowton this has been rebased :)

smowton commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,180 @@
+    mayThrowCertificateException(otherMethod)
+    or
+    not otherMethod.fromSource() and
+    otherMethod.getAnException().getType().getASupertype*() instanceof CertificatException
```
```suggestion
    otherMethod.getAnException().getType().getASupertype*() instanceof CertificateException
```

smowton commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,180 @@
+         * Regex: (_)* :
+         * some methods have underscores.
+         * Regex: (no|ignore|disable)(strictssl|ssl|verify|verification)
+         * noStrictSSL ignoreSsl
```
```suggestion
         * Examples: noStrictSSL ignoreSsl
```
(and similarly below)

smowton replied:
Please address

intrigus-lgtm replied:
Removed in 865044333e8c8556cf7d874d66434d5816408eda.

smowton commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,180 @@
+bindingset[result]
+private string getAFlagName() {
+  result
+      .regexpMatch("(?i).*(secure|disable|selfCert|selfSign|validat|verif|trust|ignore|nocertificatecheck).*")
```
Why do the flags and the sanitiser function names differ?

smowton replied:
Still outstanding ^^

smowton replied:
Please address

intrigus-lgtm replied:
Removed in 865044333e8c8556cf7d874d66434d5816408eda.

smowton commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,180 @@
+    )
+  }
+
+  override predicate isSanitizer(DataFlow::Node barrier) {
```
This might be better phrased as a side-condition on the sink, rather than tagging potentially a lot of nodes as sanitisers

smowton replied:
Still outstanding ^^

intrigus-lgtm replied:
Removed in 865044333e8c8556cf7d874d66434d5816408eda.

smowton commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,180 @@
+  )
+}
+
+/** Holds if there is flow from `node1` to `node2` either due to local flow or due to custom flow steps. */
```
Explicitly give examples of the custom flow steps

smowton replied:
Still outstanding ^^

smowton replied:
Please address

intrigus-lgtm replied:
Done in 865044333e8c8556cf7d874d66434d5816408eda.

smowton commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,180 @@
+}
+
+/** Holds if `node` is guarded by a flag that suggests an intentionally insecure feature. */
+private predicate isNodeGuardedByFlag(DataFlow::Node node) {
```
This is reimplementing the `Configuration.isBarrierGuard` concept

intrigus-lgtm replied:
So what do you suggest going forward?

smowton replied:
I'll see if I can fit this into a barrier-guard tomorrow.

smowton reviewed (commented):
How about instead of implementing custom logic for barrier guards to pick out the various "if(acceptAllCertificates) ..." cases, we treat `context.getSocketFactory` as the sink and `init(...)` as a taint propagator? That way we'll pick out cases where the `if(acceptAll) ...` logic happens in a parent function, while the flow from instantiating a bad trust manager to calling context.init happens entirely within a child function. By lengthening the passage of code we consider as the source -> sink path, we should be able to use `Configuraton.isBarrierGuard/isSanitizerGuard` instead of implementing our own `guard.controls` logic.

intrigus-lgtm commented:
> How about instead of implementing custom logic for barrier guards to pick out the various "if(acceptAllCertificates) ..." cases, we treat `context.getSocketFactory` as the sink and `init(...)` as a taint propagator? That way we'll pick out cases where the `if(acceptAll) ...` logic happens in a parent function, while the flow from instantiating a bad trust manager to calling context.init happens entirely within a child function. By lengthening the passage of code we consider as the source -> sink path, we should be able to use `Configuraton.isBarrierGuard/isSanitizerGuard` instead of implementing our own `guard.controls` logic.

How strong do you feel about this?
If you want this changed, I guess this should change as well?
https://github.com/github/codeql/blob/main/java/ql/src/Security/CWE/CWE-297/UnsafeHostnameVerification.ql#L109-L152

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.qhelp:
```diff
@@ -0,0 +1,43 @@
+<example>
+<p>
+In the first (bad) example, the <code>TrustManager</code> never throws a <code>CertificateException</code> thereby trusting any certificate.
+This allows an attacker to perform a man-in-the-middle attack.
```
```suggestion
This allows an attacker to perform a machine-in-the-middle attack.
```

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.qhelp:
```diff
@@ -0,0 +1,43 @@
+<overview>
+<p>
+If the <code>checkServerTrusted</code> method of a <code>TrustManager</code> never throws a <code>CertificateException</code> it trusts every certificate.
+This allows an attacker to perform a Man-in-the-middle attack against the application therefore breaking any security Transport Layer Security (TLS) gives.
```
```suggestion
This allows an attacker to perform a machine-in-the-middle attack against the application therefore breaking any security Transport Layer Security (TLS) gives.
```

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,180 @@
+/**
+ * @name Everything trusting `TrustManager`
+ * @description Trusting all certificates allows an attacker to perform a machine-in-the-middle attack.
```
Done in 1a33f8c202d7bd26864f1e2c957fa34ab2c640e2

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.qhelp:
```diff
@@ -0,0 +1,43 @@
+If the <code>checkServerTrusted</code> method of a <code>TrustManager</code> never throws a <code>CertificateException</code> it trusts every certificate.
+This allows an attacker to perform a Man-in-the-middle attack against the application therefore breaking any security Transport Layer Security (TLS) gives.
+
+An attack would look like this:
```
This has not been applied.
Instead it has been changed to ["An attack might look like this:"](https://github.com/intrigus-lgtm/ql/blob/2931e1f3fb4f68205c20db825d85d81f6782631b/java/ql/src/Security/CWE/CWE-297/UnsafeHostnameVerification.qhelp#L12).
Hope this is ok.

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.qhelp:
```diff
@@ -0,0 +1,43 @@
+This allows an attacker to perform a Man-in-the-middle attack against the application therefore breaking any security Transport Layer Security (TLS) gives.
+
+An attack would look like this:
+1. The program connects to <code>https://example.com</code>.
```
Applied.

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.qhelp:
```diff
@@ -0,0 +1,43 @@
+An attack would look like this:
+1. The program connects to <code>https://example.com</code>.
+2. The attacker intercepts this connection and presents a valid, self-signed certificate for <code>https://example.com</code>.
+3. Java calls the <code>checkServerTrusted</code> method to check whether it should trust the certificate.
```
Applied.

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.qhelp:
```diff
@@ -0,0 +1,43 @@
+2. The attacker intercepts this connection and presents a valid, self-signed certificate for <code>https://example.com</code>.
+3. Java calls the <code>checkServerTrusted</code> method to check whether it should trust the certificate.
+4. The <code>checkServerTrusted</code> method of your <code>TrustManager</code> does not throw a <code>CertificateException</code>.
+5. Java proceeds with the connection since your <code>TrustManager</code> implicitly trusted it by not throwing an exception.
```
Same.

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.qhelp:
```diff
@@ -0,0 +1,43 @@
+3. Java calls the <code>checkServerTrusted</code> method to check whether it should trust the certificate.
+4. The <code>checkServerTrusted</code> method of your <code>TrustManager</code> does not throw a <code>CertificateException</code>.
+5. Java proceeds with the connection since your <code>TrustManager</code> implicitly trusted it by not throwing an exception.
+6. The attacker can now read the data your program sends to <code>https://example.com</code> and/or alter its replies while the program thinks the connection is secure.
```
Currently not applied as this is the same as https://github.com/intrigus-lgtm/ql/blob/2931e1f3fb4f68205c20db825d85d81f6782631b/java/ql/src/Security/CWE/CWE-297/UnsafeHostnameVerification.qhelp#L23-L24
If you still want to see it changed, I can apply it.

smowton replied:
I see, ok let's leave this the way it is for now and when I get a moment I'll see if that query can be improved.

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,180 @@
+}
+
+/**
+ *Holds if:
```
Applied.

intrigus-lgtm commented:
@smowton I applied most of your suggestions besides to suggestions in the `.qhelp` file.
Let me know your thoughts as this section is very similar to the `InsecureHostnameVerification.qhelp` file.

Suggestions on the barrier stuff are also unresolved, as it is not yet clear how we want to proceed.

smowton commented:
OK, if you don't want to pursue that you don't have to; I'll start a benchmark run with the query as-is

smowton commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,180 @@
+ * - `m` calls a method declared to throw a `CertificateException`, but for which no source is available
+ */
+private predicate mayThrowCertificateException(Method m) {
+  exists(Stmt stmt | m.getBody().getAChild*() = stmt |
+    stmt.(ThrowStmt).getThrownExceptionType().getASupertype*() instanceof CertificateException
+  )
```
```suggestion
    m.getBody().getAChild*().(ThrowStmt).getThrownExceptionType().getASupertype*() instanceof CertificateException
```

intrigus-lgtm replied:
Applied in 865044333e8c8556cf7d874d66434d5816408eda.

smowton reviewed (commented):
There are also various unaddressed comments here. They don't all have to be applied, but please do respond to them

intrigus-lgtm commented:
> There are also various unaddressed comments here. They don't all have to be applied, but please do respond to them

Will do, but I'm currently lacking a continuous chunk of time needed for evaluating the barrier guard stuff which could make some comments outdated.

felicitymay commented:
👋🏻 I've removed myself as a reviewer so that I don't get nagged about reviewing this PR every morning 😄 

When this is ready for a docs team review, please ping the new `@github/docs-content-codeql` team.

intrigus-lgtm commented:
Sorry for the long wait.

I've experimented a bit with the query.
1. A query **without any filtering**, run only on projects that were previously identified when run on all of lgtm.com.
    557 of 557 hits.
2. A query with all filtering applied, that is, **filtering based on source/sink enclosing callable name** and **filtering based on flag guards**.
    473 of 551 hits.
3. A query with partial filtering, that is, only **filtering based on flag guards**.
    528 of 551 hits.
4. A query based on [this idea](https://github.com/github/codeql/pull/4879#pullrequestreview-570374053) with **no filtering**.
    400 of 551 hits.
5. A query based on [this idea](https://github.com/github/codeql/pull/4879#pullrequestreview-570374053) with partial filtering, that is, only **filtering based on flag guards**.
    387 of 551 hits.

With filtering on flag guards I mean code like this:
```java
if(enableHttps) { 
  sink();
}
```
Re:
1. Nothing unexpected here.
2. It would be possible to further reduce the number of hits, by improving the filtering based on source/sink names.
3. This shows a small reduction of hits. It has the big advantage that I don't have to guess the intention of the developer.
    In the case of source/sink filtering, if a method is named `disableSSLSecurity`, why is it named this way? 
    a) This method may be for external (=user controlled) consumption only. This should not be detected.
    b) The dev wanted to create a re-usable method (for their code only) and always calls it. This should be detected.
    But I'm pretty sure that it's not possible to differentiate between a) and b), because how would one determine whether something is for external consumption?
4. This shows a reduction, although there are probably sinks where the sink is not `sslContext.getSocketFactory()` but a call to another API that only receives the `sslContext`.
5. A small further reduction.

Links for the query runs have been sent to @smowton.

smowton commented:
Which version do you want to submit for consideration for a bounty?

intrigus-lgtm commented:
> Which version do you want to submit for consideration for a bounty?

The best one :D 

But I don't know what you would consider best.
Approach 2 can likely be tuned to filter most results that are likely FPs, but the downside would be that it *may* also filter TPs.
3 has more FPs by design but will likely not loose TPs.
4 & 5 don't really show a reduction of FPs. Just an overall reduction of results, I think.

smowton commented:
I don't have the time to assess 5 variants to decide which is best -- you should pick one. To assess which is doing a better job you could use the LGTM API to determine which projects differ between the different versions and take a closer look at those ones. Then you'll know whether version 4 or 5 is actually useful or whether your guess that they're just producing less results is correct.

smowton commented:
Looks like the security lab have completed their assessments regardless. Do you want to pick one of the variants above and ask them to re-evaluate, or do you want us to merge this to experimental as it is and evaluate the current version for the bounty?

intrigus-lgtm commented:
Sorry for all the delay.

I still want to get this merged into *non*-experimental.

intrigus-lgtm commented:
I think I've responded to all review comments besides https://github.com/github/codeql/pull/4879#discussion_r559494060

smowton commented:
Do you want to pick one of the variants you suggest above, or do you want me to consider this for merging as it stands?

smowton commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.java:
```diff
@@ -0,0 +1,50 @@
+		tmf.init(keyStore);
+		TrustManager[] trustManagers = tmf.getTrustManagers();
+		context.init(null, trustManagers, null); // GOOD, we are not using a custom `TrustManager` but instead have
+												// added the self-signed certificate we want to trust to the key
+												// store. Note, the `trustManagers` will **only** trust this one
+												// certificate.
```
```suggestion
		                                         // added the self-signed certificate we want to trust to the key
		                                         // store. Note, the `trustManagers` will **only** trust this one
		                                         // certificate.
```

intrigus-lgtm replied:
Done in https://github.com/github/codeql/pull/4879/commits/b50b9c0a2923b4d68c4041c82a9b67b3441e79ad

intrigus-lgtm commented:
I've chosen approach 3 in https://github.com/github/codeql/pull/4879/commits/865044333e8c8556cf7d874d66434d5816408eda:
> A query with partial filtering, that is, only filtering based on flag guards.

I do not plan further changes.

smowton commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,159 @@
+  exists(MethodAccess ma | ma.getMethod().getName() = getAFlagName() |
+    source.asExpr() = ma and
+    ma.getType() instanceof FlagType and
+    not isEqualsIgnoreCaseMethodAccess(ma)
```
This will be easier to understand if factored into `getAFlagName` -- `and result != "equalsIgnoreCase"` would do fine.

intrigus-lgtm replied:
Done in https://github.com/github/codeql/pull/4879/commits/b3f49541b508822f614fffefb653841908c143ea

smowton commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,159 @@
+}
+
+/**
+ * Holds if there is flow from `node1` to `node2` either due to local flow or due to custom flow steps:
```
```suggestion
 * Holds if there is local flow from `node1` to `node2` either due to standard data-flow steps or the following custom flow steps:
```

intrigus-lgtm replied:
Done in https://github.com/github/codeql/pull/4879/commits/b3f49541b508822f614fffefb653841908c143ea

smowton commented:
Do you think the tests in this PR are sufficient to check whether my alternative barrier-guard implementation is sufficient? If not, please add more.

intrigus-lgtm commented:
> Do you think the tests in this PR are sufficient to check whether my alternative barrier-guard implementation is sufficient? If not, please add more.

"alternative barrier-guard implementation" as in your idea from https://github.com/github/codeql/pull/4879#pullrequestreview-570374053?
If so, I don't think there are any test-cases for this.

> How about instead of implementing custom logic for barrier guards to pick out the various "if(acceptAllCertificates) ..." cases, we treat context.getSocketFactory as the sink and init(...) as a taint propagator? That way we'll pick out cases where the if(acceptAll) ... logic happens in a parent function, while the flow from instantiating a bad trust manager to calling context.init happens entirely within a child function. By lengthening the passage of code we consider as the source -> sink path, we should be able to use Configuraton.isBarrierGuard/isSanitizerGuard instead of implementing our own guard.controls logic.

smowton commented:
Because the relevant flag check is

```
  exists(VarAccess v | v.getVariable().getName() = getAFlagName() |
    source.asExpr() = v and v.getType() instanceof FlagType
  )
  or
  exists(StringLiteral s | s.getRepresentedString() = getAFlagName() | source.asExpr() = s)
  or
  exists(MethodAccess ma | ma.getMethod().getName() = getAFlagName() |
    source.asExpr() = ma and
    ma.getType() instanceof FlagType and
    not isEqualsIgnoreCaseMethodAccess(ma)
  )
```

Please ensure there is at least one test for each of:
* Guard using a string literal
* Guard using a named variable
* Guard using a method access
* The particular exclusion on equalsIgnoreCase

smowton commented:
That will put in me in a position to try using `isBarrierGuard` to replace the manually-written barrier guard and see whether it still passes the expected tests. Ideally you should also include negative tests -- ones similar to a guard, but which should be flagged because the guard does not control the source/sink, for example.

intrigus-lgtm commented:
Will try to do this until tomorrow.

intrigus-lgtm commented:
I hope the tests in https://github.com/github/codeql/pull/4879/commits/eb64e7958cd84e77dacbc4a52aeb429b0d0c7fc9 are what you had in mind in https://github.com/github/codeql/pull/4879#issuecomment-817901773.

smowton commented:
Very thorough, thanks!

smowton commented:
@intrigus-lgtm just noticed this targets non-experimental, is that intended?

smowton commented:
OK never mind the barrier guard, I can see now that this is simpler than a barrier guard -- those check for things like `if(x != 0) { ...` indicating `x` is safe to use within the `if` or `else` branch, but here we (a) don't know which branch is interesting and (b) only exclude the sink, not generally any taint propagation, so the case is sufficiently different we don't need to use BarrierGuard.

intrigus-lgtm commented:
> @intrigus-lgtm just noticed this targets non-experimental, is that intended?

Yes.

intrigus-lgtm commented:
@smowton is there anything missing from my side?
Or anything else I can/should do?
(Besides the merge conflict and the change note)

smowton commented:
Not from me; have passed back to @aschackmull because this targets non-experimental

aschackmull commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,156 @@
+ * 2. A call to an `EnvReadMethod` such as `System.getProperty` where a tainted value is used as an argument.
+ *    The return value of such a method is then tainted.
+ */
+private predicate flagFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
```
There's a lot of code here that looks directly copied from `UnsafeHostnameVerification.ql`. Please factor this out into a small library, say, `semmle.code.java.security.SecurityFlag`.

intrigus-lgtm replied:
Done in b4b9ae4

aschackmull commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,156 @@
+import DataFlow::PathGraph
+
+/**
+ * Models an insecure `X509TrustManager`.
```
```suggestion
 * An insecure `X509TrustManager`.
```

aschackmull commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,156 @@
+
+/**
+ * Holds if:
+ * - `m` may `throw` a `CertificateException`
+ * - `m` calls another method that may throw
```
Let's make it explicit that these conditions are disjunctive rather than conjunctive.
```suggestion
 * - `m` may `throw` a `CertificateException`, or
 * - `m` calls another method that may throw, or
```

aschackmull commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,156 @@
+ * - `m` calls a method declared to throw a `CertificateException`, but for which no source is available
+ */
+private predicate mayThrowCertificateException(Method m) {
+  m.getBody().getAChild*().(ThrowStmt).getThrownExceptionType().getASupertype*() instanceof
```
Use `Stmt.getEnclosingCallable()` instead of `Callable.getBody().getAChild*()`. This avoids a transitive closure operation.

intrigus-lgtm replied:
Done via https://github.com/github/codeql/pull/4879/files#r622466963

aschackmull commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,156 @@
+}
+
+/**
+ * A configuration to model the flow of a `InsecureX509TrustManager` to an `SSLContext.init` call.
```
```suggestion
 * A configuration to model the flow of an `InsecureX509TrustManager` to an `SSLContext.init` call.
```

aschackmull commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,156 @@
+/**
+ * A configuration to model the flow of a `InsecureX509TrustManager` to an `SSLContext.init` call.
+ */
+class InsecureTrustManagerConfiguration extends TaintTracking::Configuration {
```
Should this configuration be a `DataFlow::Configuration` rather than a `TaintTracking::Configuration`? It seems that we only want to track object identity here.

intrigus-lgtm replied:
@aschackmull I'm pretty sure it did not work with `DataFlow`, because the `TrustManager` flows into an array:
https://github.com/github/codeql/pull/4879/files#diff-49b72d7c1a043c36e446ab36dae514d5108fec68e33a7c19e2677ff805b5995aR3-R21
Will this be fixed with #5751 + `DataFlow`?
Or should I switch this to `DataFlow` + custom steps for arrays?

aschackmull replied:
Hmm, right, gotcha. Ideally something like this should be fixed by that PR, yes, but there are still a number of problems to work out before that PR can be merged. In particular, and related to this exact example, how should sinks be handled if the flow that reaches them is suddenly wrapped in an array.
Let's keep it as a TaintTracking config for now.

intrigus-lgtm replied:
Ok, resolving this then.

skedwards88 commented:
👋 Docs first responder here. Thanks for this PR! I've pinned it for review by the docs content team.

felicitymay commented:
👋🏻 Hi, could someone fix the conflicts so that the Qhelp preview generator can be run?

intrigus-lgtm commented:
Force-pushed to solve the merge conflicts.

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,156 @@
+ * - `m` calls a method declared to throw a `CertificateException`, but for which no source is available
+ */
+private predicate mayThrowCertificateException(Method m) {
+  m.getBody().getAChild*().(ThrowStmt).getThrownExceptionType().getASupertype*() instanceof
+    CertificateException
```
```suggestion
  exists(ThrowStmt throwStmt |
    throwStmt.getThrownExceptionType().getASupertype*() instanceof CertificateException
  |
    throwStmt.getEnclosingCallable() = m
  )
```

intrigus-lgtm commented:
@felicitymay sorry, I thought I had a script that would automatically prevent me from committing `qhelp` files with conflicts :(
But it should be fixed now :tada:

intrigus-lgtm commented:
@aschackmull I factored out the "feature-flag" library in b4b9ae4.
It's based on/inspired by https://github.com/github/codeql-go/blob/42c1116ac750daa540bfc7658c455b4d0d6b56ff/ql/src/semmle/go/security/InsecureFeatureFlag.qll

felicitymay commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.qhelp:
```diff
@@ -0,0 +1,48 @@
+
+<example>
+<p>
+In the first (bad) example, the <code>TrustManager</code> never throws a <code>CertificateException</code> thereby trusting any certificate.
```
"thereby" isn't a commonly used word, so I'd suggest:

```suggestion
In the first (bad) example, the <code>TrustManager</code> never throws a <code>CertificateException</code> and therefore implicitly trusts any certificate.
```

felicitymay commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.qhelp:
```diff
@@ -0,0 +1,48 @@
+<p>
+In the first (bad) example, the <code>TrustManager</code> never throws a <code>CertificateException</code> thereby trusting any certificate.
+This allows an attacker to perform a machine-in-the-middle attack.
+In the second (good) example, no custom <code>TrustManager</code> is used. Instead, the self-signed certificate that should be trusted
+is explicitly trusted by loading it into a <code>KeyStore</code>.
```
Suggest maybe:

```suggestion
In the second (good) example, the self-signed certificate that should be trusted
is loaded into a <code>KeyStore</code>. This explicitly defines the certificate as trusted and there is no need to create a custom <code>TrustManager</code>.
```

felicitymay commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.qhelp:
```diff
@@ -0,0 +1,48 @@
+
+<references>
+<li><a href="https://developer.android.com/training/articles/security-ssl">Android Security Guide for TLS/HTTPS</a>.</li>
+<li>OWASP: <a href="https://cwe.mitre.org/data/definitions/295.html">CWE-295</a>.</li>
```
This is a reference to the CWE Mitre site, not OWASP. However, the tag in the query for CWE-295 will ensure that a link to the CWE definition is added automatically when the help is rendered.

```suggestion
```

intrigus-lgtm replied:
Yep, normally I notice this in other queries :D

felicitymay commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.qhelp:
```diff
@@ -0,0 +1,48 @@
+</example>
+
+<references>
+<li><a href="https://developer.android.com/training/articles/security-ssl">Android Security Guide for TLS/HTTPS</a>.</li>
```
Updated to match the [style guide for links](https://github.com/github/codeql/blob/main/docs/query-help-style-guide.md#websites).

```suggestion
<li>Android Develoers:<a href="https://developer.android.com/training/articles/security-ssl">Security with HTTPS and SSL</a>.</li>
```

felicitymay commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.java:
```diff
@@ -0,0 +1,50 @@
+        // Use it with our key store that trusts our self-signed certificate
+        tmf.init(keyStore);
+        TrustManager[] trustManagers = tmf.getTrustManagers();
+        context.init(null, trustManagers, null); // GOOD, we are not using a custom `TrustManager` but instead have
+                                                // added the self-signed certificate we want to trust to the key
+                                                // store. Note, the `trustManagers` will **only** trust this one
+                                                // certificate.
```
This would probably render better in the finished query help if the comments were on separate lines. These are rather long.

intrigus-lgtm replied:
Done in c307520682c968332ec1577ed535aeecb959f736.

felicitymay reviewed (commented):
I've reviewed the user-facing text in the `InsecureTrustManager` query and made a few suggestions. Generally it's clear and I found the example of an attack particularly helpful. ✨ 

One test is failing because this query needs to have a change note. For examples, see: https://github.com/github/codeql/tree/main/java/change-notes.

intrigus-lgtm commented on java/ql/src/Security/CWE/CWE-295/InsecureTrustManager.ql:
```diff
@@ -0,0 +1,117 @@
+/**
+ * An insecure `X509TrustManager`.
+ * An `X509TrustManager` is considered insecure if it never throws a `CertificateException`
+ * thereby accepting any certificate as valid.
```
```suggestion
 * and therefore implicitly trusts any certificate as valid.
```

intrigus-lgtm commented:
@felicitymay 
- [x] suggestions applied
- [x] change note added

smowton reviewed (commented):
Looks good to me; requesting review from @aschackmull because this targets the main query set

felicitymay commented:
I'm going to move this off our team's review board as the docs team is happy with these content changes. Feel free ping us again if you need a further review, but otherwise we're happy for this to be merged when the Java CodeQL team are ready.

intrigus-lgtm commented:
@smowton _Just_ checking, did this fall through some crack?
Or is there anything that is missing and I didn't realize it?

smowton commented:
No, just @aschackmull is very busy :)

aschackmull commented:
I don't see any real reason why this shouldn't target `experimental`. We prefer to move queries from `experimental` and into the main query set at our own pace. Please move all changes such that they are confined within the `experimental` subdir.

intrigus-lgtm commented:
Although I don't really understand how this is so different from https://github.com/github/codeql/pull/4771 which has been accepted, I've moved it back to experimental...

Please review.

aschackmull commented on java/ql/src/semmle/code/java/security/SecurityFlag.qll:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Flags suggesting an optional feature, perhaps deliberately insecure.
+ */
+class SecurityFeatureFlag extends FlagKind {
```
```suggestion
private class SecurityFeatureFlag extends FlagKind {
```

aschackmull replied:
This doesn't need to be public, does it?

intrigus-lgtm replied:
Sorry, I missed that, you're right.

aschackmull commented on java/ql/src/semmle/code/java/security/SecurityFlag.qll:
```diff
@@ -0,0 +1,84 @@
+
+  /** Gets a node representing a (likely) security flag. */
+  DataFlow::Node getAFlag() {
+    exists(VarAccess v | v.getVariable().getName() = getAFlagName() |
+      result.asExpr() = v and v.getType() instanceof FlagType
+    )
+    or
+    exists(StringLiteral s | s.getRepresentedString() = getAFlagName() | result.asExpr() = s)
+    or
+    exists(MethodAccess ma | ma.getMethod().getName() = getAFlagName() |
+      result.asExpr() = ma and
+      ma.getType() instanceof FlagType
+    )
+    or
+    flagFlowStep*(getAFlag(), result)
```
This is needlessly quadratic (it will terminate in exactly 3 iterations, and iteration 3 will essentially calculate all reachable pairs of the transitive closure `flagFlowStep*`). Rewrite as follows:
```suggestion
    exists(DataFlow::Node flag |
      exists(VarAccess v | v.getVariable().getName() = getAFlagName() |
        flag.asExpr() = v and v.getType() instanceof FlagType
      )
      or
      exists(StringLiteral s | s.getRepresentedString() = getAFlagName() | flag.asExpr() = s)
      or
      exists(MethodAccess ma | ma.getMethod().getName() = getAFlagName() |
        flag.asExpr() = ma and
        ma.getType() instanceof FlagType
      )
    |
      flagFlowStep*(flag, result)
    )
```

aschackmull commented on java/ql/src/semmle/code/java/security/SecurityFlag.qll:
```diff
@@ -0,0 +1,84 @@
+  FlagKind() { any() }
+
+  /**
+   * Returns a flag name of this type.
```
```suggestion
   * Gets a flag name of this type.
```

vedantmgoyal2009 commented:
Closed #4879

smowton commented:
@intrigus-lgtm please rebase on latest `main` to make the framework coverage job happy

intrigus-lgtm commented:
@smowton rebased.

smowton commented:
@intrigus-lgtm we noticed a commit applied through the web interface (accepting @aschackmull's comments) was accidentally dropped in the rebase, and have re-applied it.

Tip for the future: if you use `--force-with-lease` instead of `--force`, `git` will warn you if the remote branch has moved in the meantime, indicating you need to `git pull` before rebasing.

intrigus-lgtm commented:
> @intrigus-lgtm we noticed a commit applied through the web interface (accepting @aschackmull's comments) was accidentally dropped in the rebase, and have re-applied it.
> 
> Tip for the future: if you use `--force-with-lease` instead of `--force`, `git` will warn you if the remote branch has moved in the meantime, indicating you need to `git pull` before rebasing.

Good catch!

I normally try to avoid this by manually checking that the remote branch has not moved in the meantime, but I forgot this time.

Thanks for letting me know about `--force-with-lease` which sounds like a less error-prone way :tada:

smowton commented:
@intrigus-lgtm looks like the tests for this query are failing due to a change in the dataflow graph layout. They seem to be `nodes` and `edges` query failures which are immaterial to the actual test result, so it should be ok to just accept the new result.

intrigus-lgtm commented:
Can you tell me which test is failing?
On 50d9ea7de27e3398e620067d97ff5f831d76ee81 I get no errors for the two tests that might have been affected by my changes.
(`Found compatible version of CodeQL CLI (version 2.5.6)`) 

```
Executing 1 tests in 1 directories.
Extracting test database in /home/sg/dev/ql/codeql-home/codeql-repo/java/ql/test/experimental/query-tests/security/CWE-295/InsecureTrustManager.
Compiling queries in /home/sg/dev/ql/codeql-home/codeql-repo/java/ql/test/experimental/query-tests/security/CWE-295/InsecureTrustManager.
Executing tests in /home/sg/dev/ql/codeql-home/codeql-repo/java/ql/test/experimental/query-tests/security/CWE-295/InsecureTrustManager.
[1/1 eval 2.4s] PASSED /home/sg/dev/ql/codeql-home/codeql-repo/java/ql/test/experimental/query-tests/security/CWE-295/InsecureTrustManager/InsecureTrustManager.qlref
All 1 tests passed.
```

```
Executing 1 tests in 1 directories.
Extracting test database in /home/sg/dev/ql/codeql-home/codeql-repo/java/ql/test/query-tests/security/CWE-297.
Compiling queries in /home/sg/dev/ql/codeql-home/codeql-repo/java/ql/test/query-tests/security/CWE-297.
Executing tests in /home/sg/dev/ql/codeql-home/codeql-repo/java/ql/test/query-tests/security/CWE-297.
[1/1 eval 2.2s] PASSED /home/sg/dev/ql/codeql-home/codeql-repo/java/ql/test/query-tests/security/CWE-297/UnsafeHostnameVerification.qlref
All 1 tests passed.
```

smowton commented:
It is indeed `experimental/query-tests/security/CWE-295/InsecureTrustManager/InsecureTrustManager.qlref`. Looks like the change was introduced yesterday -- fetch and rebase again (carefully :)) to see the difference

intrigus-lgtm commented:
Here we go again :)

duongxuantuan154 reviewed (commented):
- [ ]
